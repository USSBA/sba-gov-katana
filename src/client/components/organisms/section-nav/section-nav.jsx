import React from 'react'
import Waypoint from 'react-waypoint'
import _ from 'lodash'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import constants from '../../../services/constants.js'
import hurricaneIcon from 'assets/images/funding-programs/Funding_Programs_Icon_Disaster_white.png'
import styles from './section-nav.scss'
import { TRANSLATIONS } from '../../../translations.js'
import * as ModalActions from '../../../actions/show-modal.js'
import { determineMenuTileData, getLanguageOverride } from '../../../services/utils.js'
import { Link } from 'atoms'

class SectionNav extends React.Component {
  componentWillReceiveProps(nextProps) {
    if (nextProps.displayMobileNav) {
      let sectionNavIcon
      const sectionTitle = this.getNthLineage(-2).title
      const titleConstants = constants.sectionTitles
      if (sectionTitle === titleConstants.disasterAssistance) {
        sectionNavIcon = hurricaneIcon
      }
      this.props.actions.showMobileSectionNav(
        this.getNthLineage(-2),
        sectionNavIcon,
        this.getBacklinkUrl(this.getNthLineage(0), this.getNthLineage(-2), langCode)
      )
    }
  }

  isBusinessGuide(parentSiteMap, langCode) {
    let result
    if (langCode === 'es') {
      result = parentSiteMap.fullUrl === '/guia-de-negocios'
    } else {
      result = parentSiteMap.fullUrl === '/business-guide'
    }
    return result
  }

  getNthLineage(n) {
    return _.nth(this.props.lineage, n)
  }

  makeNavLinks(sectionSiteMap, currentPageSiteMap, langCode) {
    const navLinks = sectionSiteMap.children.map(function(item, index) {
      const titleLinkData = determineMenuTileData(langCode, item)
      let currentLinkClass = ''
      if (
        titleLinkData.fullUrl === currentPageSiteMap.fullUrl ||
        (langCode === 'es' &&
          currentPageSiteMap.spanishTranslation &&
          titleLinkData.fullUrl === currentPageSiteMap.spanishTranslation.fullUrl)
      ) {
        currentLinkClass = styles.currentNavLink
      }
      return (
        <li className={currentLinkClass} key={index}>
          <Link
            className="article-navigation-article-link-desktop"
            id={'desktop-article-link-' + index}
            to={titleLinkData.fullUrl}
          >
            {titleLinkData.title}
          </Link>
        </li>
      )
    })
    return navLinks
  }

  makeNavigationTitle(titleSiteMap, langCode) {
    let sectionTitle
    titleSiteMap.spanishTranslation && langCode === 'es'
      ? (sectionTitle = titleSiteMap.spanishTranslation.title)
      : (sectionTitle = titleSiteMap.title)

    return (
      <span id="article-navigation-title-desktop">
        <h3>{sectionTitle}</h3>
      </span>
    )
  }

  stickyFunctionTop() {
    return this.props.position === 'middle' ? styles.stickyTop : null
  }

  stickyFunctionBottom() {
    return this.props.position === 'bottom' ? styles.stickyBottom : null
  }

  getBacklinkUrl(businessGuideSiteMap, parentSiteMap, langCode) {
    let backLink
    if (businessGuideSiteMap.spanishTranslation && parentSiteMap.spanishTranslation && langCode === 'es') {
      backLink = (this.isBusinessGuide(businessGuideSiteMap, langCode)
        ? businessGuideSiteMap.spanishTranslation
        : parentSiteMap.spanishTranslation
      ).fullUrl
    } else {
      backLink = (this.isBusinessGuide(businessGuideSiteMap, langCode)
        ? businessGuideSiteMap
        : parentSiteMap
      ).fullUrl
    }
    return backLink
  }

  getBacklinkText(langCode) {
    const { allTopics, backTo } = TRANSLATIONS
    const backToText = backTo[langCode].text
    const { spanishTranslation, title } = this.getNthLineage(-2)

    const backTitleText = this.isBusinessGuide(this.getNthLineage(0), langCode)
      ? allTopics[langCode].text
      : spanishTranslation && langCode === 'es' ? spanishTranslation.title : title

    return `${backToText} ${backTitleText}`
  }

  render() {
    const langCode = getLanguageOverride()
    const navLinks = this.makeNavLinks(this.getNthLineage(-2), this.getNthLineage(-1), langCode)
    const navigationTitle = this.makeNavigationTitle(this.getNthLineage(-2), langCode)
    return (
      <div
        id="article-navigation-desktop"
        className={styles.sectionNav + ' ' + this.stickyFunctionTop() + ' ' + this.stickyFunctionBottom()}
      >
        <Waypoint topOffset="30px" onEnter={this.props.onTopEnter} />
        <Link
          id="article-navigation-back-button-desktop"
          className={styles.backLink}
          to={this.getBacklinkUrl(this.getNthLineage(0), this.getNthLineage(-2), langCode)}
        >
          {this.getBacklinkText(langCode)}
        </Link>
        {navigationTitle}
        <ul>{navLinks}</ul>
      </div>
    )
  }
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators(ModalActions, dispatch)
  }
}

export default connect(null, mapDispatchToProps)(SectionNav)
export { SectionNav }
