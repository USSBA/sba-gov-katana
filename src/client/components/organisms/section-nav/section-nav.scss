.sectionNav {
  display: none;

  @include grid-media( $sba-grid-large) {
    color: #fff;
    display: block;
    position: absolute;
    @include grid-column(3, $sba-grid-large);
  }
  @include grid-media($sba-grid-small){
    display: none;
  }

  a {
    display: block;
    font-size: 18px;
    font-weight: 300;
    line-height: 1.2;
    padding: 8px 30px 10px;
    text-decoration: none;
  }
  a.currentNavLink {
    font-weight: bold;
  }
  h2, h3 , h4 {
    text-align: left;
    padding: 0 30px;
  }
  h2{
    margin-bottom: 0;
  }
  li {
    border-top: 1px solid rgba(255, 255, 255, 0.25);
  }
  li:last-child {
    border-bottom: 1px solid rgba(255, 255, 255, 0.25);
  }
  .backLink {
    border-bottom: 1px solid rgba(255, 255, 255, 0.25);
    margin-bottom: 20px;
  }

  a.backLink{
    padding: 20px 30px 20px 55px;
    position: relative;
  }
  .backLink:before {
    content: "\f104";
    font-family: FontAwesome;
    font-size: 30px;
    margin-right: 15px;
    position: absolute;
    top: 12px;
    left: 30px;
  }
}

.stickyTop {
  position: fixed;
  top: 30px;
}

.stickyBottom {
  position: absolute;
  bottom: 518px;
}

.animateLeftNav{
  animation: slideNavIn 3s forwards;
}

@keyframes slideNavIn {
  0% { transform: translateX(100%); }
  100% { transform: translateX(0%); }
}

$section-nav-color-map: (
        "sba-blue": $sba-blue-dark,
        "byzantine": $byzantine,
        "money-green": $money-green-dark,
        "cobalt-blue": $cobalt-blue
);
$section-nav-dark-color-map: (
        "sba-blue": $sba-blue-darkest,
        "byzantine": $byzantine-dark,
        "money-green": $money-green-darkest,
        "cobalt-blue": $cobalt-blue-dark
);

@each $name, $value in $section-nav-color-map {
  @include theme($name, '.sectionNav li a, .backLink', null, #fff, rgba(255, 255, 255, 0.25));
  @include theme($name, '.sectionNav', $value, null, null);
}

@each $name, $value in $section-nav-dark-color-map {
  @include theme($name, '.currentNavLink', $value, #fff, $value);
  @include theme-hover($name, ".sectionNav li a, .sectionNav li, .backLink", $value, #fff, $value);
  @include theme-active($name, '.sectionNav li a', $value, #fff, $value);
  :global(.#{$name}) {
    .sectionNav li:hover + li {
      border-top: 1px solid $value;
    }
  }
}
