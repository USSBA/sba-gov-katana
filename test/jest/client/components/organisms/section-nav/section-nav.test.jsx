/*global expect*/

/* eslint-disable no-unused-vars,no-undef */
import React from 'react'
/*eslint-enable no-unused-vars*/
import { SectionNav } from 'organisms/section-nav/section-nav.jsx'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import _ from 'lodash'
import lineageBusinessGuide from '../../test-data/lineage-business-guide.json'
import lineageForPartners from '../../test-data/lineage-for-partners.json'

describe('SectionNav', () => {
  describe('Business Guide', () => {
    test('Renders a Third-level Section Navigation', () => {
      const component = renderer.create(
        <SectionNav lineage={lineageBusinessGuide} />
      )
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })

    test('Backlink points to first-level URL', () => {
      const modifiedBacklinkUrl = '/business-guide'
      const lineage = _.merge(_.cloneDeep(lineageBusinessGuide), [
        { fullUrl: modifiedBacklinkUrl }
      ])
      const component = shallow(<SectionNav lineage={lineage} />)
      expect(
        component.find('#article-navigation-back-button-desktop').props().url
      ).toBe(modifiedBacklinkUrl)
    })

    test("Backlink text is 'Back to all topics'", () => {
      const component = shallow(<SectionNav lineage={lineageBusinessGuide} />)
      expect(
        component.find('#article-navigation-back-button-desktop').props().text
      ).toBe('Back to all topics')
    })

    test('Navigation title uses first word of second-level as an H2 and the rest as H4', () => {
      const firstWord = 'Firstwordof'
      const otherWords = ' My Second Level Title'
      const modifiedSecondLevelTitle = `${firstWord}${otherWords}`
      const lineage = _.cloneDeep(lineageBusinessGuide)
      lineage[1].title = modifiedSecondLevelTitle
      const component = shallow(<SectionNav lineage={lineage} />)
      expect(
        component
          .find('#article-navigation-title-desktop')
          .find('h2')
          .text()
      ).toBe(firstWord)
      expect(
        component
          .find('#article-navigation-title-desktop')
          .find('h4')
          .text()
      ).toBe(otherWords)
    })
  })
  describe('For Partners', () => {
    test('Renders a Fourth-level Section Navigation', () => {
      const component = renderer.create(
        <SectionNav lineage={lineageForPartners} />
      )
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })
    test('Backlink points to third-level URL', () => {
      const modifiedBacklinkUrl = '/my/back/url'
      const lineage = _.merge(_.cloneDeep(lineageForPartners), [
        {},
        {},
        { fullUrl: modifiedBacklinkUrl }
      ])
      const component = shallow(<SectionNav lineage={lineage} />)
      expect(
        component.find('#article-navigation-back-button-desktop').props().url
      ).toBe(modifiedBacklinkUrl)
    })
    test('Backlink text is third-level title', () => {
      const modifiedTitle = 'My Third Level Title'
      const lineage = _.merge(_.cloneDeep(lineageForPartners), [
        {},
        {},
        { title: modifiedTitle }
      ])
      const component = shallow(<SectionNav lineage={lineage} />)
      expect(
        component.find('#article-navigation-back-button-desktop').props().text
      ).toBe(`Back to ${modifiedTitle}`)
    })
    test('Navigation title has all content in an h3', () => {
      const navigationTitle = 'Firstwordof My Third Level Title'
      const lineage = _.cloneDeep(lineageForPartners)
      lineage[2].title = navigationTitle
      const component = shallow(<SectionNav lineage={lineage} />)
      expect(
        component
          .find('#article-navigation-title-desktop')
          .find('h3')
          .text()
      ).toBe(navigationTitle)
    })
  })
})
