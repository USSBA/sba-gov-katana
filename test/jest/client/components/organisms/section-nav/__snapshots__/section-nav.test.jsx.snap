// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionNav Business Guide Renders a Third-level Section Navigation 1`] = `
<div
  className="sectionNav null null"
  id="article-navigation-desktop"
>
  <span
    style={
      Object {
        "fontSize": 0,
      }
    }
  />
  <a
    className="backLink"
    href="/business-guide"
    id="article-navigation-back-button-desktop"
  >
    Back to all topics
  </a>
  <span
    id="article-navigation-title-desktop"
  >
    <h2>
      Launch
    </h2>
    <h4>
       your business
    </h4>
  </span>
  <ul>
    <li>
      <a
        className="article-navigation-article-link-desktop "
        href="/business-guide/launch/pick-your-business-location-zoning-laws"
        id="desktop-article-link-0"
      >
        Pick your business location
      </a>
    </li>
    <li>
      <a
        className="article-navigation-article-link-desktop currentNavLink"
        href="/business-guide/launch/choose-business-structure-types-chart"
        id="desktop-article-link-1"
      >
        Choose a business structure
      </a>
    </li>
    <li>
      <a
        className="article-navigation-article-link-desktop "
        href="/business-guide/launch/choose-your-business-name-register"
        id="desktop-article-link-2"
      >
        Choose your business name
      </a>
    </li>
    <li>
      <a
        className="article-navigation-article-link-desktop "
        href="/business-guide/launch/register-your-business-federal-state-agency"
        id="desktop-article-link-3"
      >
        Register your business
      </a>
    </li>
    <li>
      <a
        className="article-navigation-article-link-desktop "
        href="/business-guide/launch/get-federal-state-tax-id-number-ein"
        id="desktop-article-link-4"
      >
        Get federal and state tax ID numbers
      </a>
    </li>
    <li>
      <a
        className="article-navigation-article-link-desktop "
        href="/business-guide/launch/apply-for-licenses-permits-federal-state"
        id="desktop-article-link-5"
      >
        Apply for licenses and permits
      </a>
    </li>
    <li>
      <a
        className="article-navigation-article-link-desktop "
        href="/business-guide/launch/open-business-bank-account-fees-benefits"
        id="desktop-article-link-6"
      >
        Open a business bank account
      </a>
    </li>
    <li>
      <a
        className="article-navigation-article-link-desktop "
        href="/business-guide/launch/get-business-insurance-assets-liability"
        id="desktop-article-link-7"
      >
        Get business insurance
      </a>
    </li>
  </ul>
</div>
`;

exports[`SectionNav For Partners Renders a Fourth-level Section Navigation 1`] = `
<div
  className="sectionNav null null"
  id="article-navigation-desktop"
>
  <span
    style={
      Object {
        "fontSize": 0,
      }
    }
  />
  <a
    className="backLink"
    href="/for-partners/lenders/7a-loans"
    id="article-navigation-back-button-desktop"
  >
    Back to 7(a) Loans
  </a>
  <span
    id="article-navigation-title-desktop"
  >
    <h3>
      7(a) Loans
    </h3>
  </span>
  <ul>
    <li>
      <a
        className="article-navigation-article-link-desktop "
        href="/for-partners/lenders/7a-loans/types-of-7a-loans"
        id="desktop-article-link-0"
      >
        Types of 7(a) Loans
      </a>
    </li>
    <li>
      <a
        className="article-navigation-article-link-desktop "
        href="/for-partners/lenders/7a-loans/operate-as-a-7a-lender"
        id="desktop-article-link-1"
      >
        Operate as a 7(a) Lender
      </a>
    </li>
    <li>
      <a
        className="article-navigation-article-link-desktop currentNavLink"
        href="/for-partners/lenders/7a-loans/secondary-markets"
        id="desktop-article-link-2"
      >
        Secondary Markets
      </a>
    </li>
    <li>
      <a
        className="article-navigation-article-link-desktop "
        href="/for-partners/lenders/7a-loans/training-and-courses"
        id="desktop-article-link-3"
      >
        Training and Courses
      </a>
    </li>
    <li>
      <a
        className="article-navigation-article-link-desktop "
        href="/for-partners/lenders/7a-loans/terms-and-eligibility"
        id="desktop-article-link-4"
      >
        Terms and Eligibility
      </a>
    </li>
  </ul>
</div>
`;
